src/server/frontend_wayland/wp_viewporter.h

```diff
@@ -0,0 +1,78 @@
+
+/*
+ * Copyright © Canonical Ltd.
+ *
+ * This program is free software: you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 3,
+ * as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#ifndef MIR_FRONTEND_WP_VIEWPORTER_H
+#define MIR_FRONTEND_WP_VIEWPORTER_H
+
+#include "mir/wayland/weak.h"
+#include "viewporter_wrapper.h"
+#include "wayland_wrapper.h"
+
+#include "mir/geometry/rectangle.h"
+
+#include <optional>
+
+namespace mir::frontend
+{
+class WlSurface;
+
+class WpViewporter : public wayland::Viewporter::Global
+{
+public:
+    WpViewporter(wl_display* display);
+
+private:
+    void bind(wl_resource* new_wp_viewporter) override;
+};
+
+/**
+ * Manages the wp_viewport state
+ *
+ * Threadsafety: This is a Wayland object, and should only be accessed from the Wayland thread
+ */
+class Viewport : public wayland::Viewport
+{
+public:
+    Viewport(wl_resource* new_viewport, WlSurface* surface);
+    ~Viewport() override;
+
+    /**
+     * Have the source or destination values changed since the last call to dirty()?
+     */
+    auto dirty() -> bool;
```

Comment:
Not sure "dirty" is the best name here. Obvious interpretations are "is dirty" and "make dirty", but this is neither. And the actual meaning won't be obvious reading the code that calls it. Even the comment isn't clear about what the first call does.

Maybe `unchanged_since_last_call()`?

Or, `unreported_changes()`?

Reply:
I've gone for `changed_since_last_resolve()`, and updated the implementation slightly to make that true.